core/services/pg/advisory_lock.go

```diff
@@ -3,46 +3,176 @@ package pg
+	l.conn, err = l.db.Connx(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed checking out connection from pool")
+	}
+	return nil
 }
 
-// Lock obtains exclusive session level advisory lock if available.
+// getLock obtains exclusive session level advisory lock if available.
 // It will either obtain the lock and return true, or return false if the lock cannot be acquired immediately.
-func (l *Lock) Lock(ctx context.Context) (bool, error) {
-	result := false
+func (l *advisoryLock) getLock(ctx context.Context) (locked bool, err error) {
+	l.logger.Trace("Taking advisory lock")
 	sqlQuery := "SELECT pg_try_advisory_lock($1)"
-	err := l.conn.QueryRowContext(ctx, sqlQuery, l.id).Scan(&result)
-	return result, err
+	err = l.conn.QueryRowContext(ctx, sqlQuery, l.id).Scan(&locked)
+	return locked, err
 }
 
-// Unlock releases the lock and DB connection.
-func (l *Lock) Unlock(ctx context.Context) error {
-	sqlQuery := "SELECT pg_advisory_unlock($1)"
-	_, err := l.conn.ExecContext(ctx, sqlQuery, l.id)
-	if err != nil {
-		return err
+func (l *advisoryLock) logRetry(count int) {
+	if count%1000 == 0 || (count < 1000 && count&(count-1) == 0) {
+		l.logger.Infow("Another application holds the advisory lock, waiting...", "tryCount", count)
 	}
-	// Returns the connection to the connection pool
-	return l.conn.Close()
 }
 
-// NewLock returns a Lock with *sql.Conn
-func NewLock(ctx context.Context, id int64, db *sql.DB) (Lock, error) {
-	// Obtain a connection from the DB connection pool and store it and use it for lock and unlock operations
-	conn, err := db.Conn(ctx)
-	if err != nil {
-		return Lock{}, err
+const checkAdvisoryLockStmt = `SELECT EXISTS (SELECT 1 FROM pg_locks WHERE locktype = 'advisory' AND pid = pg_backend_pid() AND (classid::bigint << 32) | objid::bigint = $1)`
+
+func (l *advisoryLock) loop() {
+	defer l.wg.Done()
+
+	ticker := time.NewTicker(utils.WithJitter(l.checkInterval))
+	defer ticker.Stop()
+
+	ctxStop, cancel := utils.ContextFromChan(l.chStop)
+	defer cancel()
+
+	for {
+		select {
+		case <-l.chStop:
+			ctx, cancel := DefaultQueryCtx()
+			err := multierr.Combine(
+				utils.JustError(l.conn.ExecContext(ctx, `SELECT pg_advisory_unlock($1)`, l.id)),
+				l.conn.Close(),
+			)
+			cancel()
+			if err != nil {
+				l.logger.Warnw("Error trying to unlock advisory lock on shutdown", "err", err)
+			}
+			return
+		case <-ticker.C:
+			var gotLock bool
+
+			ctx, cancel := context.WithTimeout(ctxStop, l.checkInterval)
+			l.logger.Trace("Checking advisory lock")
+			err := l.conn.QueryRowContext(ctx, checkAdvisoryLockStmt, l.id).Scan(&gotLock)
+			if errors.Is(err, sql.ErrConnDone) {
+				// conn was closed and advisory lock lost, try to check out a new connection and re-lock
+				l.logger.Warnw("DB connection was unexpectedly closed; checking out a new one", "err", err)
+				if err = l.checkoutConn(ctx); err != nil {
+					l.logger.Warnw("Error trying to checkout connection", "err", err)
```

Comment:
Should this branch do something like `continue` or `return`? Or was `l.getLock` meant to be in an `else` block?

Reply:
Good catch, fixed